# DX 602 Final Project

## Introduction

In this project, you will practice the skills that you have learned throughout this module with a heavy focus on building models.
Most of the problems and questions are open ended compared to your previous homeworks, and you will be asked to explain your choices.
Most of them will have a particular type of solution implied, but it is up to you to figure out the details based on what you have learned in this module.

## Instructions

Each problem asks you to perform build models, run a computation, or otherwise perform some analysis of the data, and usually answer some questions about the results.
Make sure that your question answers are well supported by your analysis and explanations; simply stating an answer without support will earn minimal points.

Notebook cells for code and text have been added for your convenience, but feel free to add additional cells.

## Example Code

You may find it helpful to refer to this GitHub repository of Jupyter notebooks for example code.

* https://github.com/bu-cds-omds/dx601-examples
* https://github.com/bu-cds-omds/dx602-examples

Any calculations demonstrated in code examples or videos may be found in these notebooks, and you are allowed to copy this example code in your homework answers.

## Submission

This project will be entirely manually graded.
However, we may rerun some or all of your code to confirm that it works as described.

### Late Policy

The normal homework late policy for OMDS does not apply to this project.
Boston University requires final grades to be submitted within 72 hours of class instruction ending, so we cannot accommodate 5 days of late submissions.

However, we have delayed the due date of this project to be substantially later than necessary given its scope, and given you more days for submission with full credit than you would have had days for submission with partial credit under the homework late policy.
The Thanksgiving holiday was also taken into account in setting the deadline.
Finally, the deadlines for DX 601 and DX 602 were coordinated to be a week apart while giving ample time for both of their projects.

## Shared Imports

For this project, you are forbidden to use modules that were not loaded in this template.
While other modules are handy in practice, modules that trivialize these problems interfere with our assessment of your own knowledge and skills.

If you believe a module covered in the course material (not live sessions) is missing, please check with your learning facilitator.

In [33]:
import math
import sys

In [34]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import scipy.stats
import sklearn.linear_model

## Problems

### Problem 1 (5 points)

Pick one of the following data sets to analyze in this project.
Load the data set, and show a random sample of 10 rows.

* [Wine Quality](https://archive.ics.uci.edu/dataset/186/wine+quality) ([PMLB - red subset only](https://github.com/EpistasisLab/pmlb/tree/master/datasets/wine_quality_red))
* [Body Fat](https://www.openml.org/search?type=data&status=active&id=560) ([PMLB](https://github.com/EpistasisLab/pmlb/tree/master/datasets/560_bodyfat))

The PMLB copies of the data are generally cleaner and recommended for this project, but the other links are provided to give you more context.
To load the data from the PMLB Github repository, navigate to the `.tsv.gz` file in GitHub and copy the link from the "Raw" button.

If the dataset has missing data, you should drop the rows with missing data before proceeding.
If the data set you choose has more than ten columns, you may limit later analysis that is requested per column to just the first ten columns.

In [35]:
# YOUR CODE HERE

# load data
wine = pd.read_csv('wine-quality-red.tsv', sep='\t')
wine

# random sample of 10 rows
wine_random_sample_10 = wine.sample(10)
wine_random_sample_10

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
728             6.4             0.570         0.02             1.8      0.067   
796             8.7             0.460         0.31             2.5      0.126   
1203            9.9             0.740         0.19             5.8      0.111   
468            11.4             0.360         0.69             2.1      0.090   
596            12.4             0.400         0.51             2.0      0.059   
114             7.8             0.560         0.19             1.8      0.104   
401             7.7             0.260         0.30             1.7      0.059   
1582            6.1             0.715         0.10             2.6      0.053   
355             6.7             0.750         0.01             2.4      0.078   
193             7.6             0.550         0.21             2.2      0.071   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
728                   4.0                  11.0  0.99700  3.46       0.68   
796                  24.0                  64.0  0.99746  3.10       0.74   
1203                 33.0                  76.0  0.99878  3.14       0.55   
468                   6.0                  21.0  1.00000  3.17       0.62   
596                   6.0                  24.0  0.99940  3.04       0.60   
114                  12.0                  47.0  0.99640  3.19       0.93   
401                  20.0                  38.0  0.99490  3.29       0.47   
1582                 13.0                  27.0  0.99362  3.57       0.50   
355                  17.0                  32.0  0.99550  3.55       0.61   
193                   7.0                  28.0  0.99640  3.28       0.55   

      alcohol  target  
728       9.5       5  
796       9.6       5  
1203      9.4       5  
468       9.2       6  
596       9.3       6  
114       9.5       5  
401      10.8       6  
1582     11.9       5  
355      12.8       6  
193       9.7       5

For this project, I chose the Wine Quality dataset. The random 10 samples are shown above.

### Problem 2 (10 points)

List all of the columns and describe them in your own words.

In [36]:
# YOUR CODE HERE

first_10_columns = wine.columns[:10]
first_10_columns

wine[first_10_columns].describe()

fixed acidity  volatile acidity  citric acid  residual sugar  \
count    1599.000000       1599.000000  1599.000000     1599.000000   
mean        8.319637          0.527821     0.270976        2.538806   
std         1.741096          0.179060     0.194801        1.409928   
min         4.600000          0.120000     0.000000        0.900000   
25%         7.100000          0.390000     0.090000        1.900000   
50%         7.900000          0.520000     0.260000        2.200000   
75%         9.200000          0.640000     0.420000        2.600000   
max        15.900000          1.580000     1.000000       15.500000   

         chlorides  free sulfur dioxide  total sulfur dioxide      density  \
count  1599.000000          1599.000000           1599.000000  1599.000000   
mean      0.087467            15.874922             46.467792     0.996747   
std       0.047065            10.460157             32.895324     0.001887   
min       0.012000             1.000000              6.000000     0.990070   
25%       0.070000             7.000000             22.000000     0.995600   
50%       0.079000            14.000000             38.000000     0.996750   
75%       0.090000            21.000000             62.000000     0.997835   
max       0.611000            72.000000            289.000000     1.003690   

                pH    sulphates  
count  1599.000000  1599.000000  
mean      3.311113     0.658149  
std       0.154386     0.169507  
min       2.740000     0.330000  
25%       3.210000     0.550000  
50%       3.310000     0.620000  
75%       3.400000     0.730000  
max       4.010000     2.000000

In [37]:
wine[first_10_columns].var()


fixed acidity              3.031416
volatile acidity           0.032062
citric acid                0.037947
residual sugar             1.987897
chlorides                  0.002215
free sulfur dioxide      109.414884
total sulfur dioxide    1082.102373
density                    0.000004
pH                         0.023835
sulphates                  0.028733
dtype: float64

1) Fixed acidity: This is a numeric column with no missing values. The values range from 4.6 to 15.9. With a variance of approximately 3.03, values vary quite significantly considering the range. This refers to the concentration of non-volatile acids in the wine, such as citric acid, that contribute to its overall taste and quality.

2) Volatile acidity: This is a numeric column with no missing values. The values range from 0.12 to 1.58. The variance is about 0.032. As the name suggests, this refers to the concentration of volatile acids in the wine. These acids contribute to a vinegary smell when present in high concentrations.

3) Citric acid: This is a numeric column with no missing values. The values range from 0 to 1 with a variance of about 0.038, similar to volatile acidity. This refers to the concentration of citric acid in the wine which contributes to its taste. With its relation to fixed acidity, a fair assumption is that fixed acidity and citric acid are positively correlated.

4) Residual sugar: This is a numeric column with no missing values. The values range from 0.9 to 15.5 with a variance of about 1.99. This refers to the amounnt of sugar left after fermentation. Naturally, the higher the amount, the sweeter the wine.

5) Chlorides: This is a numeric column with no missing values. The values range from 0.012 to 0.611 with a relatively low variance of 0.002215. Chlorides, which are the mineral salts in the wine, contribute to its level of saltiness.

6) Free sulfur dioxide: This is a numeric column with no missing values. The values range from 1 to 72 with a variance of approximately 109, which is a significantly high variance compared to other columns. We can assume this may rank high if you run a principal component analysis on this dataset. This refers to the antioxidant and antimicrobial agent present in wines that act as preservative.

7) Total sulfur dioxide: This is a numeric column with no missing values. The values range from 6 to 289 with a variance of 1806 which is even higher than free sulfur oxide. The range is also higher/more spread out than other columns. The max value might also be an outlier in this case. This refers to the overall amount of sulfur dioxide in the wine, including the free SO2 and the bound SO2.

8) Density: This is a numeric column with no missing values. The values range from 0.99 to 1.00369 with a relatively low variance of 0.000004. This extremly low variance reveals that wine density hardly vary from wine to wine.

9) pH: This is a numeric column with no missing values. The values range from 2.74 to 4.01 with a variance of about 0.024, meaning the pH level of the different wine samples vary within a reasonable range.

10) Sulphates: This is a numeric column with no missing values. The values range from 0.33 to 2 with a variance of about 0.029. Similar to pH level, the values don't vary that much. This refers to the concentration of sulphates in the wine which is a natural preservative.


### Problem 3 (50 points)

Perform an exploratory analysis of the data set.
After your exploratory analysis, pick 3 individual charts that you the think were particularly interesting.
Repeat those charts separately from your original analysis, and after each of those charts, explain what you thought was noteworthy.

In [38]:
# YOUR CODE HERE

YOUR ANSWERS HERE

### Problem 4 (5 points)

Plot the correlation matrix of the numeric columns in the data set.
Which pair of different columns were highlighted as the most correlated?

In [39]:
# YOUR CODE HERE

YOUR ANSWERS HERE

### Problem 5 (10 points)

Pick three different regression model classes to try in problem 6 from the scikit-learn documentation.
For each class, provide a link to the scikit-learn documentation, and a link to another web page describing how that kind of model works.
The second link should not be from scikit-learn, but Wikipedia is acceptable.
You do not need to understand the methods at this time, but it is good to be comfortable researching them.

In [40]:
# YOUR CODE HERE

YOUR ANSWERS HERE

### Problem 6 (50 points)

Build three different regression models using the entire data set.
Plot the actual target vs the predicted values for each in one chart.
Compute the L2 and L1 losses for each of them.
You may use any regression class provided provided by scikit-learn, and you may reuse one class as long as you change its parameters enough to see different results.

In [41]:
# YOUR CODE HERE

YOUR ANSWERS HERE

### Problem 7 (30 points)

Use 5-fold cross-validation to repeat building the same three kinds of regression models. Compare the L2 losses predicted by cross-validation against the L2 losses training against the whole data set. (The difference is likely from overfitting in the latter.)

In [42]:
# YOUR CODE HERE

YOUR ANSWERS HERE

### Problem 8 (25 points)

Build three different regression models as in problem 6, but preprocess the data so that each column has mean zero and standard deviation one first.
For full credit, use a scikit-learn pipeline for each model.
For each model, compare the L2 losses -- which of them performed differently from your results in problem 6?

(This process will be covered in week 13.)

In [43]:
# YOUR CODE HERE

YOUR ANSWERS HERE

### Problem 9 (5 points)

A colleague suggests that you find better models by repeatedly building decision trees with random depth limits.
They say that trying 1000 such models will likely find an improvement as long as you use cross validation.
Give a one sentence response to this suggestion. 

In [44]:
# YOUR CODE HERE

YOUR ANSWERS HERE

### Problem 10 (10 points)

Pick a best model from all the models that you built and otherwise described in this project.
Explain how you picked it, including what criteria you chose, and how the other models compared by that criteria.
As much as possible, justify that problem in the context of the original data set. 

In [45]:
# YOUR CODE HERE

YOUR ANSWERS HERE